platform/storage/zephyr/nvs/src/mender-storage.c

```diff
@@ -180,34 +180,100 @@ mender_storage_delete_authentication_keys(void) {
 }
 
 mender_err_t
-mender_storage_set_deployment_data(char *deployment_data) {
+mender_storage_set_deployment_data(cJSON *deployment_data) {
 
     assert(NULL != deployment_data);
 
-    /* Write deployment data */
-    if (nvs_write(&mender_storage_nvs_handle, MENDER_STORAGE_NVS_DEPLOYMENT_DATA, deployment_data, strlen(deployment_data) + 1) < 0) {
+    /* Add/update version number. We always want to store the deployment data in
+     * the format of the latest and greatest version. By updating the version
+     * number here and validating the data below, we make sure to enforce this.
+     */
+    cJSON *item;
+    if (NULL == (item = cJSON_CreateNumber((double)MENDER_STORAGE_DEPLOYMENT_DATA_VERSION))) {
+        mender_log_error("Unable to allocate memory");
+        return MENDER_FAIL;
+    }
+    if (NULL == cJSON_GetObjectItem(deployment_data, "version")) {
+        if (!cJSON_AddItemToObject(deployment_data, "version", item)) {
+            mender_log_error("Unable to allocate memory");
+            cJSON_Delete(item);
+            return MENDER_FAIL;
+        }
+    } else {
+        /* One would expect cJSON_AddItemToObject above to replace the value on
+         * already existing keys. However, it creates another entry with a
+         * duplicate key, which is invalid JSON. So we have to be careful. */
+        if (!cJSON_ReplaceItemInObject(deployment_data, "version", item)) {
+            mender_log_error("Unable to allocate memory");
+            cJSON_Delete(item);
+            return MENDER_FAIL;
+        }
+    }
+
+    /* Validate deployment data */
+    if (!mender_utils_validate_deployment_data(deployment_data, MENDER_STORAGE_DEPLOYMENT_DATA_VERSION)) {
+        mender_log_error("Invalid deployment data");
+        return MENDER_FAIL;
+    }
+
+    /* Compose JSON string */
+    char *json_str;
+    if (NULL == (json_str = cJSON_PrintUnformatted(deployment_data))) {
+        mender_log_error("Unable to compose deployment data");
+        return MENDER_FAIL;
+    }
```

Comment:
Everything above this line in this function is platform-agnostic, right? Can we put the code into a common place, please?